File: LarpakeServer/Models/DatabaseModels/Attendance.cs
using LarpakeServer.Data.TypeHandlers;

namespace LarpakeServer.Models.DatabaseModels;

public class Attendance
{
    public required Guid UserId { get; set; }

    [SqlColumn("larpake_event_id")]
    public required long LarpakeTaskId { get; set; }
    public Guid? CompletionId { get; set; } = null;
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public string? QrCodeKey { get; set; } = null;
    public DateTime? KeyInvalidAt { get; set; } = null;

    public Completion? Completion { get; set; } = null;

    public static Attendance From(long eventId, Guid userId)
    {
        return new Attendance
        {
            UserId = userId,
            LarpakeTaskId = eventId,
        };
    }
}

File: Frontend/src/models/attendance.ts
import { LarpakeTask } from "./larpake";
import { User } from "./user";

export type Completion = {
    id: string;
    signerId: string;
    signatureId: string | null;
    completedAt: Date;
    createdAt: Date;
    updatedAt: Date;
};

export type AttendanceKey = {
    key: string;
    qrCodeKey: string;
    keyInvalidAt: Date;
};

export type Attendance = {
    userId: string;
    larpakeTaskId: number;
    completed: Completion | null;
    createdAt: Date;
    updatedAt: Date;
    key: AttendanceKey | null;
};

export type FatAttendance = {
    userId: string;
    larpakeTaskId: number;
    completed: Completion | null;
    createdAt: Date;
    updatedAt: Date;
    key: AttendanceKey | null;
    user: User;
    task: LarpakeTask;
};
